Add Ledger API Test Tool docs into DAML docs (#587)
digital-asset/daml: docs/source/tools/ledger-api-test-tool/index.rst

.. Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

Ledger API Test Tool
####################

The Ledger API Test Tool is a command line tool for testing the correctness of
implementations of the :doc:`Ledger API
</app-dev/ledger-api-introduction/index>`, i.e. DAML ledgers. For example, it

> will show you if there is consistency or conformance problem with your

Reviewer: ```suggestion
will show you if there are consistency or conformance problems with your
```

Author: Done

implementation.

Its intended audience are developers of DAML ledgers, who are using the
:doc:`DAML Ledger Implementation Kit </ledger-integration-kit/index>` to develop
a DAML ledger on top of their distributed-ledger or database of choice.

Use this tool to verify if your Ledger API endpoint conforms to the :doc:`DA
Ledger Model </concepts/ledger-model/index>`.

Please note that currently the tool is in ALPHA status:

- ALPHA: contains a subset of the tests we use internally to test our ledger
  implementations.
- BETA: will be reached once we extended the coverage to include the full set of
  our test as per <https://github.com/digital-asset/daml/issues/146>.

Downloading the tool
====================

TODO: To be defined in <https://github.com/digital-asset/daml/issues/142>.

Extracting ``.dar`` file required to run the tests
======================================================

Before you can run the Ledger API test tool on your ledger, you need to load a
specific set of DAML templates onto your ledger.

#. To obtain the corresponding ``.dar`` file, run:

>    ``ledger-api-test-tool -x``

Reviewer: `-x` is a pretty cryptic flag. Does it have a fuller, clearer name you can use instead?

Author: Replaced with the full name `--extract` here and below

.. Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

> .. _ledger-api-test-tool:

Reviewer: As before - do you need this reference?

> Ledger API Test Tool

Reviewer: Just my opinion. You refer to it in lowercase later.

```suggestion
Ledger API test tool
```

Reviewer: I don't think things have to be capitalised just because they're names, but people seem to like it - they think it makes it more of a Thing somehow. It's not really necessary, but totally up to you.

Reviewer 2: I'd prefer keeping the approach of capitalizing the names

####################

> The Ledger API Test Tool is a command line tool for testing the correctness of

Reviewer: What does "test the correctness of" mean? A concrete example would make this clearer - something like "For example, it will show you if there is x problem with your implementation"

> The Ledger API Test Tool is a command line tool for testing the correctness of

Reviewer: ```suggestion
The Ledger API test tool is a command line tool for testing the correctness of
```

ledger implementations based on DAML and the :doc:`Ledger API
</app-dev/ledger-api-introduction/index>`. It’s useful for ledger implementation
developers, who are using the DAML Ledger Implementation Kit to develop a DAML
Ledger on top of their distributed-ledger or database of choice.

> Use this tool to verify if the Ledger API endpoint conforms to the :doc:`DA

Reviewer: your ledger API endpoint?

Ledger Model </concepts/ledger-model/index>`.

> 

Reviewer 2: Add a paragraph clarifying the status of the tool, which is ALPHA.
- ALPHA: contains a subset of the tests we use internally to test our ledger implementations
- BETA: will be reached once we extended the coverage to include the full set of our test (ADD LINK TO ISSUE)

Tag issue with DLIK label and add to backlog for DLIK: https://github.com/digital-asset/daml/milestone/5

Reviewer: +1

> .. contents:: On this page:

Reviewer: Not needed with the new doc site - already has a TOC on the right.

  :local:

Downloading the tool
====================

> TODO

Reviewer: Isn't the page a bit useless without this bit?

Extracting :code:`.dar` file required to run the tests
======================================================

Before you can run the Ledger API test tool on your ledger, you need to load a
specific set of DAML templates onto your ledger.

> #. Obtain the corresponding :code:`.dar` file with:

Reviewer: Why are you using `:code:` rather than just double backticks?

> #. Obtain the corresponding :code:`.dar` file with:

Reviewer: ```suggestion
#. To get the corresponding :code:`.dar` file, run:
```

> #. Obtain the corresponding :code:`.dar` file with:

Reviewer: Do I need to run this anywhere in particular?

>    ``ledger-api-test-tool -x``

Reviewer: I'd add a line after this that says "This creates a file ``SemanticTests.dar``" Then, in the next point, you can just say "Load ``SemanticTests.dar`` into your ledger"

#. Load the file :code:`SemanticTests.dar` created in the current directory into your
   Ledger.

Running the tool against a custom Ledger API endpoint
=====================================================

Run this command to test your Ledger API endpoint exposed at host :code:`<host>` and
at a port :code:`<port>`:

    ``ledger-api-test-tool -h <host> -p <port>``

For example

    ``ledger-api-test-tool -h localhost -p 6865``

Exploring options the tool provides
===================================

Run the tool with :code:`--help` flag to obtain the list of options the tool provides:

>    .. code-block:: console

Reviewer: Or just tell them to run the command, and don't include the output?

     $ ledger-api-test-tool --help

     The Ledger API Test Tool is a command line tool for testing the correctness of
     ledger implementations based on DAML and Ledger API.
     Usage: ledger-api-test-tool [options]

       --help                   prints this usage text
       -p, --target-port <value>

>                                 Ledger API server port. Defaults to 6865.

Reviewer: Obviously this is a comment for the output itself, but this should really be "Server port of the Ledger API endpoint to test" or something

>        -h, --host <value>       Ledger API server host. Defaults to localhost.

Reviewer: Same comment as above - specify it's the endpoint you want to test

>        --must-fail              One or more of the scenario tests must fail. Defaults to false.

Reviewer: This isn't really sufficient. I'd say something like "Use this flag if you expect one or more or the scenario tests to fail", and tell them what they can expect to happen if they use this. How does it change the output? What's going to happen?

Author: Done

>        -r, --reset              Perform a ledger reset before running the tests. Defaults to false.

Reviewer: Why would I want to use this? A sentence explaining what it's for would help.

Author: Done

>        -x, --extract            Extract the testing archive files and exit.

Reviewer: I don't understand what this will do. Is this after running all the tests?

Author: Done

> Try out the Ledger API Test Tool against DAML Sandbox

Reviewer: ```suggestion
Try out the Ledger API test tool against DAML Sandbox
```

=====================================================

> To run the tool against [DAML

Reviewer: ```suggestion
If you wanted to test out the tool, you can run it against [DAML
```

Author: Done

> Sandbox](https://docs.daml.com/tools/sandbox.html), run:

Reviewer: ```suggestion
Sandbox](https://docs.daml.com/tools/sandbox.html). To do this:
```

Author: Done

> Sandbox](https://docs.daml.com/tools/sandbox.html), run:

Reviewer: +1. I don't think this even works in rst?

> Sandbox](https://docs.daml.com/tools/sandbox.html), run:

Author: Done

   .. code-block:: console

     ledger-api-test-tool -x
     da sandbox -- SemanticTests.dar
     ledger-api-test-tool

> This should always succeed! This is useful if you do not have yet a custom

Reviewer: +1, with my comment above that adds the why at the top.

> This should always succeed! This is useful if you do not have yet a custom

Author: Done

Ledger API endpoint.

> The DAML Sandbox starts at by default at :code:`localhost:6865`

Reviewer 2: use `Note that the DAML Sandbox` to clarify the relation of this paragraph to preceding text.

Reviewer: Or start with the why - "You don't need to supply the hosts and ports arguments, because the Ledger API test tool defaults to using ``localhost:6865``, which the Sandbox uses by default."

Author: Done

> The DAML Sandbox starts at by default at :code:`localhost:6865`

Author: Used @bethaitman suggestion

and the Ledger API Test Tool uses this as the default endpoint to test, hence
hosts and ports command line arguments can be omitted.

Testing your tool from continuous integration pipelines
=======================================================

> To test your ledger in your CI pipelines, execute it as part of your pipeline:

Reviewer 2: To test your ledger in our CI pipeline, run the `ledger-api-test` tool as part of your pipeline. You can use the `--port` option to avoid clashes of port allocations.

   .. code-block:: console

     $ ledger-api-test-tool
     Running 10 scenarios against localhost:6865...
     Testing scenario: Test:timeTravel
     Testing scenario: Test:authorization_rule1_failure
     Testing scenario: Test:authorization_success2_delegation
     Testing scenario: Test:authorization_rule2_failure
     Testing scenario: Test:startsAtEpoch
     Testing scenario: Test:authorization_success1
     Testing scenario: Test:privacy_projections1
     Testing scenario: Test:consistency_doublespend1
     Testing scenario: Test:consistency_doublespend3
     Testing scenario: Test:consistency_doublespend2
     All scenarios completed.
     $ echo $?

>      0

Reviewer 2: This code sample shows the output of the tool on the console. I'd expect that to fit better to the Sandbox example. Again there is the problem that it goes out of date quite quickly, which is why I'd recommend to not even include the output; or only with a caveat.

Reviewer: Right - I'm not sure that the output helps the user task here.

Author: The test output is not dependent on the server. I've tweaked the commands

The tool is tailored to be used in CI pipelines: as customary, when the tests
succeed, it will produce minimal output and return the success exit code.

Using the tool with a known-to-be-faulty Ledger API implementation
==================================================================

> To force the tool to always return success exit code, use :code:`--must-fail` flag:

Reviewer 2: The `must-fail` option was introduced for a different purpose: it should allow checking that the tool actually does successfully detect a failure.

A name that better captures this behaviour is `internal-expect-at-least-one-failure`. This option is not meant to be used by consumers.

Suggestion:
- explain that with the `must-fail` option the tool returns success on at least one failure, and fails if there no failure
- add explanation that we use this internally to test that the tool does not always return true
- add link to an issue for specifying the names (or patterns of names) of the tests that are expected to fail, which is the generalization that is actually useful for DAML ledger developers

Reviewer: Simon's explanation is very helpful here - it really helps to capture the purpose. 

That explains why we need it internally. Before you've implemented the feature to specify the tests that will fail, what's the value of this for external users?

Author: I think `Using the tool with a known-to-be-faulty Ledger API implementation` captures the value of this option for the external users.

@meiersi-da, I do not understand what you mean in 
```
add link to an issue for specifying the names (or patterns of names) of the tests that are expected to fail, which is the generalization that is actually useful for DAML ledger developers
```

I've tweaked the section.

    ``ledger-api-test-tool --must-fail -h localhost -p 6865``

This is useful during development of a DAML Ledger implementation, when tool

> needs to be used against a known-to-be-faulty implementation (e.g. in CI).

Reviewer: But why is it useful then? What will it allow you to do in that situation?

Author: See above

Reviewer: This doesn't quite seem sufficient to me. I guess I would have expected an option like this to return interesting information about the failures, rather than just crashing out. But maybe that's not what it does?

Author: It will always print out information about test failures. Explained it here and in the section about testing a custom API endpoint.
